pylint/checkers/refactoring/len_checker.py

```diff
@@ -85,11 +85,26 @@ def visit_call(self, node):
             parent = node.parent
             while isinstance(parent, astroid.BoolOp):
                 parent = parent.parent
-
             # we're finally out of any nested boolean operations so check if
             # this len() call is part of a test condition
             if _is_test_condition(node, parent):
-                self.add_message("len-as-condition", node=node)
+                try:
+                    instance = next(node.args[0].infer())
+                except astroid.InferenceError:
+                    self.add_message("len-as-condition", node=node)
+                    return
```

Comment:
I don't know much about the internals of astroid's infer, but what's the reasoning to emitting the error if the value can't be inferred? I'm assuming that astroid would probably less likely to be able to infer the type of an imported library, but that could be wrong

Reply:
It actually happen for list comprehension, dict comprehension, set comprehension and generators like range. It's not very elegant nor explicit..

Reply:
I don't know exactly what to think about it.
The implementation you propose is quite powerfull but it will emit a message as soon as the len argument cannot be inferred. I'm not confident about the fact it will occur only for comprehensions and generator...
 
Maybe we could try to filter comprehension before trying to infer. Something along the lines:
```
# we're finally out of any nested boolean operations so check if
# this len() call is part of a test condition
if _is_test_condition(node, parent):
    len_arg = node.args[0]
    # First, the node is a generator or comprehesion as in len([x for x in ...])
    if isinstance(len_arg, (astroid.ListComp, astroid.SetComp, astroid.DictComp, astroid.GeneratorExp):
        self.add_message("len-as-condition", node=node)
        return
    # Or the node is a name as in len(foo) and we look for name assignment and analyze it
    if isinstance(len_arg, astroid.Name):
        lookup for name assignment and check if it is a Comprehension or Generator
    # Last case i see is the node is a call to a function as in len(my_func(my_arg)) but it that case we should be able to check whether the result is a Generator
   if isinstance(len_arg, astroid.Call) and astroid.GeneratorExpr in len_arg.inferred():
       self.add_message("len-as-condition", node=node)
...
```
Note that i'm not found of this implementation neither because it uses a lot of isinstance calls...
What do you think about it?
Maybe we can ask @AWhetter advice...

Reply:
I agree, it's very heavy. I'm searching for a better way than checking all the code of the class in search of a __bool__ function. If there isn't one, maybe not raising a message for anything that is not a list, set, dict, range, comprehension or generator has a better yield in term of result versus calculation to do ?

Reply:
I think doing the checks @hippo91 suggested might even end up being faster than calling `.infer()`. If we're worried about the performance that much, we could do some performance comparisons and see which is faster.
My preference is to do the checks because we're way less likely to do raise false positives, and I think they're more annoying than pylint running slowly. At least this section of code will get called only when `len()` is calling in the source code being analysed.

Reply:
Doing it like @hippo91 suggested is fine, my implementation wasn't efficient at all.